src/dm/DispatchManagerActions.cc

```diff
@@ -2256,3 +2455,751 @@ int DispatchManager::live_updateconf(std::unique_ptr<VirtualMachine> vm,
 
 /* -------------------------------------------------------------------------- */
 /* -------------------------------------------------------------------------- */
+
+int DispatchManager::attach_sg(int vid, int nicid, int sgid,
+        const RequestAttributes& ra, string& error_str)
+{
+    ostringstream oss;
+
+    bool is_update = false;
+
+    // -------------------------------------------------------------------------
+    // Check action consistency:
+    // - Object exists VM, NIC and SG
+    // - State is compatible with operation
+    // -------------------------------------------------------------------------
+    if ( sgpool->exist(sgid) < 0 )
+    {
+        oss << "Could not attach SG " << sgid << " to VM, SG does not exist";
+        error_str = oss.str();
+
+        return -1;
+    }
+
+    if ( auto vm = vmpool->get(vid) )
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This snippet may contain an accidental use of incorrect operation. Make sure that the operation being performed is in fact what you intended.